File: ForBusiness/NetBeans/NCDataOnlineService/src/ncdataonlineservice/Sincronizador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ncdataonlineservice;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URLEncoder;
import java.security.KeyManagementException;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.config.Registry;
import org.apache.http.config.RegistryBuilder;
import org.apache.http.conn.socket.ConnectionSocketFactory;
import org.apache.http.conn.socket.PlainConnectionSocketFactory;
import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.ssl.SSLContextBuilder;
import org.apache.http.ssl.TrustStrategy;
import org.codehaus.jackson.map.ObjectMapper;
import org.codehaus.jackson.type.TypeReference;

/**
 *
 * @author tml
 */
public class Sincronizador extends Thread{
    
    static ConfigData config;
    static String configName;
    static String configFileName;
    static String configLogName;
    static Map tmp;
    static Settings settings;
    static Sincronizador first;
    static HashMap<String, Timer> timers;
    
    Sincronizador() {
        try {
            config = new ConfigData();
            configName = "";
            configFileName = "";
            configLogName = "";
            tmp = new HashMap<String,String>();
            settings = new Settings(tmp, "settings", true, "78x3s65a6de9d241", "d2c7cb7s5a74a48c");
            timers = new HashMap<String,Timer>();
        } catch (IOException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    public void run(){
        try {
            this.configLogName = Util.GetDirectoryPath() + Util.CleanString(this.configName) + ".log";
            this.configFileName = Util.GetDirectoryPath() + Util.CleanString(this.configName) + ".config";
            
            Log.Write("Windows Service do Sincronizador de Dados para NCDataOnline a arrancar", this.configLogName);
            this.UpdateConfig();
            this.StartTimers();
        } catch (IOException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    public void RunQuery(String name) 
    {
        System.out.println(name);
        try
        {
            this.UpdateConfig();
            
            if (this.config.query.get(name).get("activoQuery").equals("True"))
            {
                if (this.config.query.get(name).get("conexao").equals("") || !this.config.conexao.containsKey(this.config.query.get(name).get("conexao")))
                {
                    System.out.println("RunQuery necessita uma conexão: " + this.config.query.get("conexao") + " não é uma conexão válida.");
                    throw new IllegalStateException("RunQuery necessita uma conexão: " + this.config.query.get("conexao") + " não é uma conexão válida.");
                }
                this.DoQuery(name);
            }
        }
        catch (Exception ex2) {
            this.ExceptionMessage(ex2);
            if (!this.config.geral.get("debug").equals("True")) {
                this.timers.get("query_"+name).cancel();
            }
        }
    }
    
    private void ExceptionMessage(Exception ex)  {
        try {
            Log.Write("Ocorreu uma excepção:" + System.getProperty("line.separator") + System.getProperty("line.separator") + System.getProperty("line.separator") + ex.getMessage(), this.configLogName);
        } catch (IOException ex1) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex1);
        }
    }
    
    public void ReadDataFile(String file, HashMap<String, String> dados)  {
        String text = "";
        File f = new File(file);
        
        if (f.exists()) {
            try {
                text = Util.ReadAllText(file);
            } catch (IOException ex) {
                Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
        if (!text.equals("")) {
            try {
                TypeReference<HashMap<String,Object>> typeRef = new TypeReference<HashMap<String,Object>>() {};
                dados = new ObjectMapper().readValue(text, typeRef);
            } catch (IOException ex) {
                Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
    }
    
    public void DoQuery(String name)  { 
        String text = this.config.query.get(name).get("conexao");
        ResultSet rs = null;
        String connectionString = "";
        try {
            if (this.config.conexao.get(text).get("tipo_basedados").equals("SQLite")) {
                connectionString = SQL.MakeSQLiteConnectionString(this.config.conexao.get(text).get("bd"));
                rs = SQL.ExecutarSQL("SQLite", connectionString, this.config.query.get(name).get("sql"));
            }
            else if (this.config.conexao.get(text).get("tipo_basedados").equals("MSSQL")) {
                connectionString = SQL.MakeMSSQLConnectionString(this.config.conexao.get(text).get("server"), this.config.conexao.get(text).get("bd"), this.config.conexao.get(text).get("user"), this.config.conexao.get(text).get("password"));
                rs = SQL.ExecutarSQL("MSSQL", connectionString, this.config.query.get(name).get("sql"));
            }

            ArrayList<ArrayList<String>> al = Util.RS2AL(rs);

            int rowcount = al.size();
            int colcount = al.get(0).size();
            
            if( rowcount>0 ) {
                HashMap dictionary = new HashMap<String, Object>();

                String text2 = this.config.query.get(name).get("etiqueta");
                Integer num = 50;
                Integer num2 = 10;
                ArrayList<HashMap<String, Object>> list = new ArrayList<HashMap<String, Object>>();
                Integer num3 = 1;

                //cabecalhos
                ArrayList<String> list2 = new ArrayList<String>();
                while ( num3 <= colcount && num3 <= num2) {
                    String text3 = this.config.query.get(name).get( "val" + num3.toString() );
                    if (text3.equals(""))
                    {
                        break;
                    }
                    list2.add(text3);
                    num3++;
                }

                //linhas
                int num4 = 0;
                Iterator it = al.iterator();
                while (it.hasNext() && colcount > 0 && num4 < num) {
                    LinkedHashMap<String, Object> dictionary2 = new LinkedHashMap<String, Object>();
                    ArrayList cur_al = (ArrayList)it.next();

                    for (String temp : list2) {
                        dictionary2.put(temp, cur_al.get(list2.indexOf(temp)));
                    }
                    
                    //System.out.println(new PrettyPrintingMap<String, Object>(dictionary2));
                    
                    list.add(dictionary2);
                    num4++;

                    it.remove();
                }

                if (!dictionary.containsKey(text2))
                {
                    dictionary.put(text2, null);
                }
                
                dictionary.put(text2, list);
                LinkedHashMap<String, Object> obj = new LinkedHashMap<String, Object>();
                obj.put(text2, dictionary.get(text2));

                String json = new ObjectMapper().writeValueAsString(obj);
                
                String url_json = URLEncoder.encode(json, "UTF-8");
                this.PostPush(name, url_json);
            }
        }
        catch (SQLException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        } catch (IOException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    public void PostPush(String name, String json) 
    {
        try {
            if (this.config.geral.get("hash").equals("") || this.config.geral.get("key").equals("") || this.config.query.get(name).get("destinoidQuery").equals(""))
            {
                System.out.println("PostPush faltam dados necessários para efectuar o Push.");
                throw new IllegalStateException("PostPush faltam dados necessários para efectuar o Push.");
            }
            
            HttpClientBuilder b = HttpClientBuilder.create();
            SSLContext sslContext = new SSLContextBuilder().loadTrustMaterial(null, new TrustStrategy() {
                public boolean isTrusted(X509Certificate[] arg0, String arg1) throws CertificateException {
                    return true;
                }
            }).build();
            b.setSslcontext( sslContext);
            HostnameVerifier hostnameVerifier = SSLConnectionSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER;
            SSLConnectionSocketFactory sslSocketFactory = new SSLConnectionSocketFactory(sslContext, hostnameVerifier);
            Registry<ConnectionSocketFactory> socketFactoryRegistry = RegistryBuilder.<ConnectionSocketFactory>create()
                    .register("http", PlainConnectionSocketFactory.getSocketFactory())
                    .register("https", sslSocketFactory)
                    .build();
            PoolingHttpClientConnectionManager connMgr = new PoolingHttpClientConnectionManager( socketFactoryRegistry);
            b.setConnectionManager( connMgr);
            
            HttpClient client = b.build();
            
            String address = this.settings.GetSettingFromFile("PostPushURL").replace("[%hash%]", this.config.geral.get("hash").toString());
            
            HttpPost post = new HttpPost(address);
            String output = "";
            try {
                List<NameValuePair> data = new ArrayList<NameValuePair>(3);
                
                data.add(new BasicNameValuePair("api_key", this.config.geral.get("key").toString()));
                data.add(new BasicNameValuePair("slot", this.config.query.get(name).get("destinoidQuery")));
                data.add(new BasicNameValuePair("data", json));
                
                post.setEntity(new UrlEncodedFormEntity(data));
                
                HttpResponse response = client.execute(post);
                BufferedReader rd = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));
                String line = "";
                while ((line = rd.readLine()) != null) {
                    output = output + line;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            
            TypeReference<HashMap<String,String>> typeRef = new TypeReference<HashMap<String,String>>() {};
            HashMap response = new ObjectMapper().readValue(output, typeRef);
            
            if (! Boolean.parseBoolean(response.get("success").toString())) {
                Log.Write("Problemas no PostPush: " + response.get("error_message").toString(), this.configLogName);
                Log.Write("Output: " + output, this.configLogName);
            }
        } catch (NoSuchAlgorithmException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        } catch (KeyStoreException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        } catch (KeyManagementException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        } catch (IOException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    public String WriteDataFile(String file, HashMap<String, Object> dados) 
    {
        try {
            String text = new ObjectMapper().writeValueAsString(dados);
            if (!file.equals(""))
            {
                Util.WriteAllText(file, text);
            }
            return text;
        } catch (IOException ex) {
            Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
        }
        return "";
    }
    
    private Timer SetQueryTimer(String name) 
    {
        Integer num = 15;
        
        if (!this.config.query.get(name).get("intervaloQuery").contains(",") && this.config.query.get(name).get("intervaloQuery").contains("."))
        {
            try {
                num = Integer.parseInt(this.config.query.get(name).get("intervaloQuery").replace(",", "."));
            }
            catch(Exception ex) {
                try {
                    Log.Write("Query \"" + name + "\" tem um intervalo que não pode ser convertido para um número", this.configLogName);
                    num = 15;
                } catch (IOException ex1) {
                    Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex1);
                }
            }
        }
        else {
            try {
                num = Integer.parseInt(this.config.query.get(name).get("intervaloQuery"));
            }
            catch(Exception ex) {
                try {
                    Log.Write("Query \"" + name + "\" tem um intervalo que não pode ser convertido para um número", this.configLogName);
                    num = 15;
                } catch (IOException ex1) {
                    Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex1);
                }
            }
        }
        
        class MyTimerTask extends TimerTask  {
            Sincronizador param;

            public MyTimerTask(Sincronizador param) {
                this.param = param;
            }

            @Override
            public void run() {
                this.param.RunQuery(name);
            }
       }
        
        Timer timer = new Timer();
        timer.scheduleAtFixedRate(new MyTimerTask(this), num*60*1000, num*60*1000);
        return timer;
    }
    
    public void StartTimers() {
        try
        {
            Iterator it = this.config.query.entrySet().iterator();
            
            while (it.hasNext()) {
                Map.Entry current = (Map.Entry)it.next();

                if (this.config.geral.get("debug") == "True") {
                    this.RunQuery( current.getKey().toString() );
                }
                else {
                    this.timers.put("query_" + current.getKey(), SetQueryTimer(current.getKey().toString()));
                }

                it.remove();
            }
        }
        catch (Exception ex)
        {
            this.ExceptionMessage(ex);
        }
    }
    
    public void Stop()
    {
        try
        {
            Log.Write("Windows Service do Sincronizador de Dados para NCDataOnline a parar", this.configLogName);
            for (Map.Entry<String, Timer> entry : this.timers.entrySet()) {
                entry.getValue().cancel();
            }
            this.timers.clear();
            Log.Write("Windows Service do Sincronizador de Dados para NCDataOnline parou", this.configLogName);
        }
        catch (Exception ex)
        {
            this.ExceptionMessage(ex);
        }
    }
    
    public void StartSinc(String name)
    {
        try
        {
            this.first = new Sincronizador();
            this.first.configName = name;
            this.first.start();
        }
        catch (Exception ex)
        {
            this.ExceptionMessage(ex);
            this.Stop();
        }
    }
    
    public void UpdateConfig() 
    {
        File file = new File(this.configFileName);
        if (!file.exists()) {
            try {
                System.out.println("Não existe ficheiro de configuração: " + this.configFileName);
                Log.Write("Não existe ficheiro de configuração: " + this.configFileName, this.configLogName);
                return;
            } catch (IOException ex) {
                Logger.getLogger(Sincronizador.class.getName()).log(Level.SEVERE, null, ex);
            }
        }

        this.LoadConfig();
    }
    
    public void LoadConfig()
    {
        try
        {
            File file = new File(this.configFileName);
            if (file.exists()) {
                String text = Util.ReadAllText(this.configFileName);
                boolean flag = true;
                flag = this.settings.GetSettingFromFile("UsesCrypt", flag);
                if (flag) {
                    text = StringCipher.Decrypt(text, "78x3s65a6dc9d2x1", "d2c7cb525a74a48c");
                }

                this.config = new ObjectMapper().readValue(text, ConfigData.class);
            }
        }
        catch (Exception ex)
        {
        }
    }
}

File: ForBusiness/VisualStudio/OLD/NCDataOnline/02/NCDataOnlineService/NCDataOnlineService/Sincronizador.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.IO;
using System.Net;
using System.Collections.Specialized;
using System.Data.SqlClient;

namespace NCDataOnlineService
{
    public class Sincronizador
    {
        private ConfigData config = new ConfigData();
        private string configName = string.Empty;
        private string configFileName = string.Empty;
        private string configLogName = string.Empty;
        private static object key = new object();
        private Settings settings = new Settings(new Dictionary<string, string>(), "settings", true, "78x3s65a6de9d241", "d2c7cb7s5a74a48c");
        private Thread first;
        private Dictionary<string, System.Timers.Timer> timers = new Dictionary<string, System.Timers.Timer>();
        public void Start(string name)
        {
            try
            {
                this.configName = name;
                this.configLogName = Util.GetDirectoryPath() + Util.CleanString(this.configName) + ".log";
                this.configFileName = Util.GetDirectoryPath() + Util.CleanString(this.configName) + ".config";
                Log.Write("Windows Service do Sincronizador de Dados para NCDataOnline a arrancar", this.configLogName);
                this.UpdateConfig();
                this.first = new Thread((ThreadStart)delegate
                {
                    this.StartTimers();
                });
                this.first.IsBackground = true;
                this.first.Start();
            }
            catch (Exception ex)
            {
                this.ExceptionMessage(ex);
                this.Stop();
            }
        }
        public void Stop()
        {
            try
            {
                lock (Sincronizador.key)
                {
                    Log.Write("Windows Service do Sincronizador de Dados para NCDataOnline a parar", this.configLogName);
                    List<string> list = this.timers.Keys.ToList<string>();
                    foreach (string current in list)
                    {
                        if (this.timers[current] != null)
                        {
                            this.timers[current].Stop();
                            this.timers[current].Close();
                            this.timers[current] = null;
                        }
                    }
                    this.timers.Clear();
                    Log.Write("Windows Service do Sincronizador de Dados para NCDataOnline parou", this.configLogName);
                }
            }
            catch (Exception ex)
            {
                this.ExceptionMessage(ex);
            }
        }
        private void UpdateConfig()
        {
            if (!File.Exists(this.configFileName))
            {
                throw new Exception("Não existe ficheiro de configuração: " + this.configFileName);
            }
            DateTime lastWriteTime = File.GetLastWriteTime(this.configFileName);
            DateTime t = default(DateTime);
            DateTime.TryParse(this.settings.GetSettingFromFile("ConfigModifiedDate"), out t);
            if (t < lastWriteTime)
            {
                this.LoadConfig();
                this.PostRegcheck();
                this.settings.SetSettingToFile("ConfigModifiedDate", lastWriteTime.AddSeconds(1.0).ToString(), false);
                return;
            }
            if (this.config.geral.Count == 0)
            {
                this.LoadConfig();
            }
        }
        private void ExceptionMessage(Exception ex)
        {
            Log.Write(string.Concat(new string[]
			{
				"Ocorreu uma excepção:",
				Environment.NewLine,
				ex.Message,
				Environment.NewLine,
				Environment.NewLine,
				ex.StackTrace
			}), this.configLogName);
        }
        private void StartTimers()
        {
            lock (Sincronizador.key)
            {
                try
                {
                    foreach (KeyValuePair<string, Dictionary<string, string>> current in this.config.query)
                    {
                        if (this.config.geral["debug"] == "True")
                        {
                            this.RunQuery(current.Key);
                        }
                        else
                        {
                            this.timers.Add("query_" + current.Key, this.SetQueryTimer(current.Key));
                        }
                    }
                }
                catch (Exception ex)
                {
                    this.ExceptionMessage(ex);
                }
            }
        }
        private System.Timers.Timer SetQueryTimer(string name)
        {
            double num = 15.0;
            if (!this.config.query[name]["intervaloQuery"].Contains(",") && this.config.query[name]["intervaloQuery"].Contains("."))
            {
                this.config.query[name]["intervaloQuery"] = this.config.query[name]["intervaloQuery"].Replace(".", ",");
            }
            if (!double.TryParse(this.config.query[name]["intervaloQuery"], out num))
            {
                Log.Write("Query \"" + name + "\" tem um intervalo que não pode ser convertido para um número", this.configLogName);
                num = 15.0;
            }
            System.Timers.Timer timer = new System.Timers.Timer();
            timer.Interval = num * 60.0 * 1000.0;
            timer.Elapsed += delegate
            {
                this.RunQuery(name);
            };
            timer.Start();
            return timer;
        }
        private void LoadConfig()
        {
            try
            {
                if (File.Exists(this.configFileName))
                {
                    string text = Util.ReadAllText(this.configFileName, 60, 1000);
                    bool flag = true;
                    this.settings.GetSettingFromFile("UsesCrypt", out flag);
                    if (flag)
                    {
                        text = StringCipher.Decrypt(text, "78x3s65a6dc9d2x1", "d2c7cb525a74a48c");
                    }
                    this.config = Json.Deserialize<ConfigData>(text);
                }
            }
            catch (Exception ex)
            {
                this.ExceptionMessage(ex);
            }
        }
        private void PostRegcheck()
        {
            using (WebClient webClient = new WebClient())
            {
                NameValueCollection nameValueCollection = new NameValueCollection();
                nameValueCollection["api_key"] = this.config.geral["key"];
                nameValueCollection["register_check"] = Convert.ToBase64String(Encoding.ASCII.GetBytes(Util.ReadAllText(this.configFileName, 60, 1000)));
                string address = this.settings.GetSetting("PostRegcheckURL").Replace("[%hash%]", this.config.geral["hash"]);
                ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
                byte[] bytes = webClient.UploadValues(address, "POST", nameValueCollection);
                string @string = Encoding.ASCII.GetString(bytes);
                Dictionary<string, string> dictionary = Json.Deserialize<Dictionary<string, string>>(@string);
                if (!bool.Parse(dictionary["success"]))
                {
                    throw new Exception("Problemas no PostRegcheck: " + dictionary["error_message"]);
                }
            }
        }
        private void PostPush(string name, string json)
        {
            if (this.config.geral["hash"] == string.Empty || this.config.geral["key"] == string.Empty || this.config.query[name]["destinoidQuery"] == string.Empty)
            {
                throw new Exception("PostPush faltam dados necessários para efectuar o Push.");
            }
            using (WebClient webClient = new WebClient())
            {
                NameValueCollection nameValueCollection = new NameValueCollection();
                nameValueCollection["api_key"] = this.config.geral["key"];
                nameValueCollection["slot"] = this.config.query[name]["destinoidQuery"];
                nameValueCollection["data"] = json;
                string address = this.settings.GetSettingFromFile("PostPushURL").Replace("[%hash%]", this.config.geral["hash"]);
                byte[] bytes = webClient.UploadValues(address, "POST", nameValueCollection);
                ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
                string @string = Encoding.ASCII.GetString(bytes);
                Dictionary<string, string> dictionary = Json.Deserialize<Dictionary<string, string>>(@string);
                if (!bool.Parse(dictionary["success"]))
                {
                    Log.Write("Problemas no PostPush: " + dictionary["error_message"], this.configLogName);
                }
            }
        }
        private void RunQuery(string name)
        {
            lock (Sincronizador.key)
            {
                try
                {
                    this.UpdateConfig();
                    if (this.config.query[name]["activoQuery"] == "True")
                    {
                        if (this.config.query[name]["conexao"] == string.Empty || !this.config.conexao.ContainsKey(this.config.query[name]["conexao"]))
                        {
                            throw new Exception("RunQuery necessita uma conexão: " + this.config.query[name]["conexao"] + " não é uma conexão válida.");
                        }
                        this.DoQuery(name);
                    }
                }
                catch (SqlException ex)
                {
                    this.ExceptionMessage(ex);
                }
                catch (Exception ex2)
                {
                    this.ExceptionMessage(ex2);
                    if (this.config.geral["debug"] != "True")
                    {
                        this.timers[name].Stop();
                        this.timers[name].Close();
                    }
                }
            }
        }
        private void DoQuery(string name)
        {
            string text = this.config.query[name]["conexao"];
            string connectionString = SQL.MakeMSSQLConnectionString(this.config.conexao[text]["user"], this.config.conexao[text]["password"], this.config.conexao[text]["server"], (this.config.conexao[text]["trusted"] == "True") ? "yes" : "no", this.config.conexao[text]["bd"], "30");
            using (SqlConnection sqlConnection = new SqlConnection(connectionString))
            {
                sqlConnection.Open();
                SqlCommand sqlCommand = new SqlCommand(this.config.query[name]["sql"], sqlConnection);
                using (SqlDataReader sqlDataReader = sqlCommand.ExecuteReader())
                {
                    if (sqlDataReader.HasRows)
                    {
                        Dictionary<string, object> dictionary = new Dictionary<string, object>();
                        this.ReadDataFile(this.config.query[name]["ficheirosaida"], ref dictionary);
                        string text2 = this.config.query[name]["etiqueta"];
                        int num = 50;
                        int num2 = 10;
                        List<Dictionary<string, object>> list = new List<Dictionary<string, object>>();
                        int num3 = 1;
                        List<string> list2 = new List<string>();
                        while (num3 <= sqlDataReader.FieldCount && num3 <= num2)
                        {
                            string text3 = this.config.query[name]["val" + num3.ToString()];
                            if (!(text3 != string.Empty))
                            {
                                break;
                            }
                            list2.Add(text3);
                            num3++;
                        }
                        int num4 = 0;
                        while (sqlDataReader.Read() && sqlDataReader.FieldCount > 0 && num4 < num)
                        {
                            Dictionary<string, object> dictionary2 = new Dictionary<string, object>();
                            foreach (string current in list2)
                            {
                                dictionary2.Add(current, sqlDataReader[list2.IndexOf(current)]);
                            }
                            list.Add(dictionary2);
                            num4++;
                        }
                        if (!dictionary.ContainsKey(text2))
                        {
                            dictionary.Add(text2, null);
                        }
                        if ((this.config.query[name].ContainsKey("resultadoQuery") && this.config.query[name]["resultadoQuery"] == "Vários registos") || (!this.config.query[name].ContainsKey("resultadoQuery") && list.Count > 1))
                        {
                            dictionary[text2] = list;
                        }
                        else if ((this.config.query[name].ContainsKey("resultadoQuery") && this.config.query[name]["resultadoQuery"] == "Apenas um registo") || (!this.config.query[name].ContainsKey("resultadoQuery") && list.Count == 1))
                        {
                            dictionary[text2] = list[0];
                        }
                        else
                        {
                            dictionary[text2] = new Dictionary<string, object>();
                        }
                        this.WriteDataFile(this.config.query[name]["ficheirosaida"], dictionary);
                        Dictionary<string, object> obj = new Dictionary<string, object>
						{

							{
								text2,
								dictionary[text2]
							}
						};
                        string json = Json.Serialize<Dictionary<string, object>>(obj, this.config.geral["debug"] == "True");
                        this.PostPush(name, json);
                    }
                }
            }
        }
        private void ReadDataFile(string file, ref Dictionary<string, object> dados)
        {
            string text = string.Empty;
            if (File.Exists(file))
            {
                text = Util.ReadAllText(file, 60, 1000);
            }
            if (text != string.Empty)
            {
                dados = Json.Deserialize<Dictionary<string, object>>(text);
            }
        }
        private string WriteDataFile(string file, Dictionary<string, object> dados)
        {
            string text = Json.Serialize<Dictionary<string, object>>(dados, this.config.geral["debug"] == "True");
            if (file != string.Empty)
            {
                Util.WriteAllText(file, text, 60, 1000);
            }
            return text;
        }
    }
}
